relay-server/src/actors/project_cache.rs

```diff
@@ -77,8 +79,20 @@ impl ProjectCache {
         let eviction_start = Instant::now();
         let delta = 2 * self.config.project_cache_expiry() + self.config.project_grace_period();
 
-        self.projects
-            .retain(|_, entry| entry.last_updated_at() + delta > eviction_start);
+        let config = &self.config;
+        let garbage = &self.garbage_disposal;
+        self.projects.retain(|key, entry| {
+            let keep = entry.last_updated_at() + delta > eviction_start;
+            if !keep {
+                // Useless creation of placeholder project here.
+                // This would not be necessary with nightly `drain_filter`.
+                let mut placeholder = Project::new(*key, config.clone());
+                std::mem::swap(entry, &mut placeholder);
```

Comment:
Given this becomes quite ugly now, what if we use `hashbrown` and its [`drain_filter`](https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.drain_filter) for this?